### PR TITLE
BAL01 - Swap and Flash Loan Fee Change

### PR DESCRIPTION
**Pull request by Hyodar**
# Swap and Flash Loan Fee Change Bot

## Types of changes

- [x] New Bot
- [ ] Fix a bot error
- [ ] Update a bot logic/test/documentation
- [ ] Code style update (formatting, renaming)
- [ ] Other (please describe):

## New Bot PR readiness

- [x] `npm install` command works (i.e. package-lock is updated properly)
- [x] The `name` and `description` fields of `package.json` describe the bot properly
- [x] All the libraries listed in `package.json` are being used
- [x] All the development libraries are in `devDependencies` section of `package.json`
- [x] The bot README is updated
- [x] The bot README sufficiently describes what the bot does
- [x] All the alerts that the bot emits are listed in the README
- [x] The txn/block examples listed in the README generate the expected findings
- [x] Code is easy to follow and/or properly commented
- [x] Bot contains tests
- [x] All tests pass when running `npm test`
- [x] Tests are not making real network calls (i.e. network interaction is properly mocked)
- [x] Tests cover all the paths (i.e. negative & positive test cases)
- [x] When the tests complete, no warning/errors are shown
- [x] `npm start` works without errors when executed using a jsonRpcUrl for each of the networks specified in the README
- [x] The networks specified in the README and the networks are listed in the `package.json` are the same
- [x] The code does what the README describes
- [x] Network calls are properly being cached (if duplicated calls can occur)
- [x] A performance review has been done (e.g. `Promise.all` is used whenever possible)
- [x] The findings fields are filled out correctly
- [x] Findings `protocol` is specified (if the bot is for a specific one)
- [x] `metadata` of the findings contains all the information needed.
- [x] The SDK methods are being used appropriately (e.g. `filterLog`, `filterFunction`, `getEthersProvider`)
- [x] `ABI`s match the contracts used / `args` used from `LogDescriptions` exists in the `ABI`s
- [x] The bot follows Forta recommended [best practices](https://docs.forta.network/en/latest/best-practices/)
- [x] Bot can recover from failed error calls when they can fail (e.g. errors are being caught on calls that can fail)
- [x] `repository` is added in the `package.json`

### CODE REVIEW

File: Balancer-Bots/Swap-FlashLoan-Fee-Change/src/agent.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  ];
+
+  return async (blockEvent: BlockEvent): Promise<Finding[]> => {
+    const logs = (
+      await provider.getLogs({
+        address: networkManager.get("protocolFeesCollectorAddress"),
+        fromBlock: blockEvent.blockNumber,
+        toBlock: blockEvent.blockNumber,
+      })
+    ).filter((log) => sighashes.includes(log.topics[0]));
```
Comment (e1Ru1o) [inline Balancer-Bots/Swap-FlashLoan-Fee-Change/src/agent.ts:39]: please include the topics in the filter:
```suggestion
    const logs = (
      await provider.getLogs({
        address: networkManager.get("protocolFeesCollectorAddress"),
        fromBlock: blockEvent.blockNumber,
        toBlock: blockEvent.blockNumber,
        topics: [sighashes],
      })
    );
```

Comment (Hyodar) [reply Balancer-Bots/Swap-FlashLoan-Fee-Change/src/agent.ts:39]: Good catch, I forgot to change this in this one too when I saw there could be multiple options for a topic. Thanks!

File: Balancer-Bots/Swap-FlashLoan-Fee-Change/src/agent.spec.ts [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    const blockEvent = new TestBlockEvent().setNumber(0);
+
+    mockProvider.addLogs([
+      getLog(0, PROTOCOL_FEES_COLLECTOR_ADDRESS, "Swap", ethers.BigNumber.from("100000000000000")),
+    ]);
+
+    expect(await handleBlock(blockEvent)).toStrictEqual([createSwapFeeFinding("0.01")]);
+  });
+
+  it("should generate a finding when a SwapFeePercentageChanged event is detected", async () => {
```
Comment (0xtaf) [inline Balancer-Bots/Swap-FlashLoan-Fee-Change/src/agent.spec.ts:136]:
```suggestion
  it("should generate a finding when a FlashLoanFeePercentageChanged event is detected", async () => {
```

**Review by 0xtaf — CHANGES_REQUESTED** [review]
nothing code related. just some changes required for readme and one test name.
`npm run format` also edits 2 blocks

**Pull request merged**